### PR TITLE
Joe at openai/summarize with controllable detail

### PR DESCRIPTION
**Pull request by joe-at-openai**
## Summary

The objective of this notebook is to demonstrate how to summarize large documents with a controllable level of detail. If you give a GPT model the task of summarizing a long document (e.g. 10k or more tokens), you'll tend to get back a relatively short summary that isn't proportional to the length of the document. For instance, a summary of a 20k token document will not be twice as long as a summary of a 10k token document. One way we can fix this is to split our document up into pieces, and produce a summary piecewise. After many queries to a GPT model, the full summary can be reconstructed. By controlling the number of text chunks and their sizes, we can ultimately control the level of detail in the output.

## Motivation

I hope it's self evident!

## For new content

When contributing new content, read through our [contribution guidelines](https://github.com/openai/openai-cookbook/blob/main/CONTRIBUTING.md), and mark the following action items as completed:

- [x] I have added a new entry in [registry.yaml](https://github.com/openai/openai-cookbook/blob/main/registry.yaml) (and, optionally, in [authors.yaml](https://github.com/openai/openai-cookbook/blob/main/authors.yaml)) so that my content renders on the cookbook website.
- [x] I have conducted a self-review of my content based on the [contribution guidelines](https://github.com/openai/openai-cookbook/blob/main/CONTRIBUTING.md#rubric):
  - [x] Relevance: This content is related to building with OpenAI technologies and is useful to others.
  - [x] Uniqueness: I have searched for related examples in the OpenAI Cookbook, and verified that my content offers new insights or unique information compared to existing documentation.
  - [x] Spelling and Grammar: I have checked for spelling or grammatical mistakes.
  - [x] Clarity: I have done a final read-through and verified that my submission is well-organized and easy to understand.
  - [x] Correctness: The information I include is correct and all of my code executes successfully.
  - [x] Completeness: I have explained everything fully, including all necessary references and citations.

We will rate each of these areas on a scale from 1 to 4, and will only accept contributions that score 3 or higher on all areas. Refer to our [contribution guidelines](https://github.com/openai/openai-cookbook/blob/main/CONTRIBUTING.md) for more details.

### CODE REVIEW

**Review by shyamal-anadkat — APPROVED** [review]
lgtm

File: examples/Summarizing_with_controllable_detail.ipynb [hunk 1]
```diff
@@ -0,0 +1,657 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Summarizing with Controllable Detail"
```
Comment (ibigio) [inline examples/Summarizing_with_controllable_detail.ipynb:7]: nit: maybe "summarization" or "how to summarize..."?

File: examples/Summarizing_with_controllable_detail.ipynb [hunk 2]
```diff
@@ -0,0 +1,657 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The objective of this notebook is to demonstrate how to summarize large documents with a controllable level of detail. If you give a GPT model the task of summarizing a long document (e.g. 10k or more tokens), you'll tend to get back a relatively short summary that isn't proportional to the length of the document. For instance, a summary of a 20k token document will not be twice as long as a summary of a 10k token document. One way we can fix this is to split our document up into pieces, and produce a summary piecewise. After many queries to a GPT model, the full summary can be reconstructed. By controlling the number of text chunks and their sizes, we can ultimately control the level of detail in the output."
```
Comment (ibigio) [inline examples/Summarizing_with_controllable_detail.ipynb:14]: nit: first line is awesome and punchy - put rest of paragraph in separate text block (or add two newlines)

File: examples/Summarizing_with_controllable_detail.ipynb [hunk 3]
```diff
@@ -0,0 +1,657 @@
...
+    "    combined_chunks, _, dropped_chunk_count = combine_chunks_with_no_minimum(\n",
+    "        chunks, max_tokens, chunk_delimiter=delimiter, add_ellipsis_for_overflow=True\n",
+    "    )\n",
+    "    if dropped_chunk_count > 0:\n",
+    "        print(f\"warning: {dropped_chunk_count} chunks were dropped due to overflow\")\n",
+    "    combined_chunks = [f\"{chunk}{delimiter}\" for chunk in combined_chunks]\n",
+    "    return combined_chunks\n",
+    "\n",
+    "\n",
+    "def combine_chunks_with_no_minimum(\n",
```
Comment (ibigio) [inline examples/Summarizing_with_controllable_detail.ipynb:111]: Might be useful to introduce these functions at a high level, and maybe also how they fit together. (So a skimming reader can still follow along).

**Review by ibigio — COMMENTED** [review]
| Criteria     | Description                                                                                         | Score |
| ------------ | --------------------------------------------------------------------------------------------------- | ----- |
| Relevance    | Is the content related to building with OpenAI technologies? Is it useful to others?                |    4   |
| Uniqueness   | Does the content offer new insights or unique information compared to existing documentation?       |   4    |
| Clarity      | Is the language easy to understand? Are things well-explained? Is the title clear?                  |    2   |
| Correctness  | Are the facts, code snippets, and examples correct and reliable? Does everything execute correctly? |     4  |
| Conciseness  | Is the content concise? Are all details necessary? Can it be made shorter?                          |     4  |
| Completeness | Is the content thorough and detailed? Are there things that weren’t explained fully?                |    4   |
| Grammar      | Are there grammatical or spelling errors present?                                                   |    4   |

This is a super cool cookbook! Really like the methodology and going over the examples – mostly could benefit from some high-level explanations of the concepts you're demonstrating (and commenting) with code. People are going to love this talk!

**Pull request merged**